core/src/main/java/io/undertow/conduits/DeflatingStreamSinkConduit.java

```diff
@@ -129,21 +132,42 @@ public int write(final ByteBuffer src) throws IOException {
                     return 0;
                 }
             }
-            byte[] data = new byte[src.remaining()];
-            src.get(data);
-            preDeflate(data);
-            deflater.setInput(data);
-            Connectors.updateResponseBytesSent(exchange, 0 - data.length);
+            ByteBuffer arrayBuffer;
+            ByteBufferPool pool = exchange.getConnection().getByteBufferPool().getArrayBackedPool();
+            if (inputBytes <= pool.getBufferSize()) {
+                if (inputBuffer == null) {
+                    inputBuffer = pool.allocate();
+                } else {
+                    inputBuffer.getBuffer().clear();
+                }
+                arrayBuffer = inputBuffer.getBuffer();
+                arrayBuffer.put(src);
+                arrayBuffer.flip();
+            } else {
+                freeInputBuffer();
+                byte[] data = new byte[inputBytes];
+                src.get(data);
+                arrayBuffer = ByteBuffer.wrap(data);
+            }
+
+            preDeflate(
+                    arrayBuffer.array(),
+                    arrayBuffer.arrayOffset() + arrayBuffer.position(),
+                    arrayBuffer.arrayOffset() + arrayBuffer.limit() - arrayBuffer.position());
+            deflater.setInput(
+                    arrayBuffer.array(),
+                    arrayBuffer.arrayOffset() + arrayBuffer.position(),
+                    arrayBuffer.arrayOffset() + arrayBuffer.limit() - arrayBuffer.position());
+            Connectors.updateResponseBytesSent(exchange, 0 - inputBytes);
             deflateData(false);
-            return data.length;
+            return inputBytes;
         } catch (IOException | RuntimeException | Error e) {
             freeBuffer();
             throw e;
         }
     }
 
-    protected void preDeflate(byte[] data) {
-
+    protected void preDeflate(byte[] data, int offset, int length) {
```

Comment:
This could be considered a break if anything besides GzipStreamSinkConduit extends this class.